### PR TITLE
Save image link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [ ] The landing page links to the image link submission form.
- [ ] The link is entered through a form.
- [ ] After the form is saved, the link is persisted in the database.
- [ ] After the form is submitted, the user is redirected to a page displaying
  the image.
- [ ] I cannot successfully save an image with an invalid URL.
- [ ] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by jnogales319**
Closes #2 

https://trello.com/c/ugTfwgBJ/18-issue-2-save-image-link

### CODE REVIEW

File: app/controllers/image_controller.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class ImageController < ActionController::Base
+  def index; end
+end
```
Comment (blakeherrington) [inline app/controllers/image_controller.rb]: Here's this image controller we were talking about - I think we can probably remove this from this commit, do agree? Since we know that this gets deleted later?

File: app/views/welcome/index.html.erb [hunk 2]
```diff
@@ -1 +1 @@
-<h1>Bienvenidos a localhost:3000!</h1>
+<%= link_to('Add an image +',{:action => 'main', :controller => 'image'}) %>
```
Comment (blakeherrington) [inline app/views/welcome/index.html.erb]: It looks maybe you started working on this in this commit and then took a different direction later - right now this link wouldn't work right? 

It might be cleaner to just go ahead and remove this file from this commit too - then we just have a clean commit that introduces the simple form dependency.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get '/images/new'
+    assert_response :success
+
+    puts response.body
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: We can probably remove this `puts`

File: test/controllers/welcome_controller_test.rb [hunk 4]
```diff
@@ -6,6 +6,7 @@ class WelcomeControllerTest < ActionDispatch::IntegrationTest
     assert_response :success
 
     puts response.body
-    assert_select 'h1', count: 1, text: 'Bienvenidos a localhost:3000!'
+    # assert_select "a:match('href', ?)", '/images/new', count: 1
```
Comment (blakeherrington) [inline test/controllers/welcome_controller_test.rb]: Did you mean to leave this commented line here?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -7,6 +7,6 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
 
     puts response.body
     # rails inserts a hidden form input, see https://www.seqmedia.com/2011/02/18/rails-forms-automatically-insert-a-hidden-utf8-field/
-    assert_select 'form input', 2
+    assert_select 'form input', 3
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Instead of working around the hidden input, we could make this test more specific by adding classes to the url input and submit button (see simple form documentation). This also would be a better test - because we are checking for specific fields, not just that there are inputs.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -21,13 +21,14 @@ def create
...
-    end
+  # Use callbacks to share common setup or constraints between actions.
+  def set_image
+    @image = Image.find(params[:id])
+  end
+
+  # Only allow a trusted parameter "white list" through.
+  def image_params
+    params.require(:image).permit(:url)
+  end
```
Comment (blakeherrington) [inline app/controllers/images_controller.rb]: Any reason why you changed the order of these two functions in this commit?

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+class ImagesController < ApplicationController
+  before_action :set_image, only: [:show]
```
Comment (blakeherrington) [inline app/controllers/images_controller.rb]: I know this is scaffolding, but since this before_action is only used in `show`, I actually think the code would be more readable if you eliminated `set_image` all together and just inlined that code into your show action.

What do you think?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -1,12 +1,28 @@
 require 'test_helper'
 
 class ImagesControllerTest < ActionDispatch::IntegrationTest
-  test 'should get new' do
-    get '/images/new'
+  def test_new__basics_intact
+    get new_image_url
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Do these changes belong in this commit? Seems unrelated to adding the validation.

File: config/routes.rb [hunk 9]
```diff
@@ -1,5 +1,5 @@
 Rails.application.routes.draw do
-  resources :images, only: [:show, :new, :create]
+  resources :images, only: %i[show new create]
```
Comment (blakeherrington) [inline config/routes.rb]: This looks like a rubocop change - seems like this might belong in another commit as well. 😄

File: app/controllers/image_controller.rb (hunk 1 above)
Comment (jnogales319) [reply app/controllers/image_controller.rb]: I agree, I dropped this controller shortly after making the PR.

File: test/controllers/welcome_controller_test.rb (hunk 4 above)
Comment (jnogales319) [reply test/controllers/welcome_controller_test.rb]: It was left in case I needed an alternate format to satisfy rubocop, but should be gone now.

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (jnogales319) [reply app/controllers/images_controller.rb]: I think the change here is that I decreased the indents per rubocop's suggestion. The order looks the same unless I'm looking at the wrong thing.

File: app/controllers/images_controller.rb (hunk 7 above)
Comment (jnogales319) [reply app/controllers/images_controller.rb]: That sounds reasonable.

File: app/views/welcome/index.html.erb (hunk 2 above)
Comment (jnogales319) [reply app/views/welcome/index.html.erb]: Agreed. Taken care of.

File: test/controllers/images_controller_test.rb (hunk 8 above)
Comment (jnogales319) [reply test/controllers/images_controller_test.rb]: Wiped away when redoing commit history.

File: test/models/image_test.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+    image.valid?
+
+    assert_equal 'must be a valid image url.', image.errors[:url].first
+  end
+
+  def test_url_is_valid_image__non_image_url_invalid
+    image = Image.new(url: 'https://www.google.com/')
+
+    image.valid?
```
Comment (blakeherrington) [inline test/models/image_test.rb]: Minor, but I'm noticing here we could have done `refute_predicate image, :valid?` here. Up to you whether or not you want to make the change.

File: app/views/images/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<p id="notice"><%= notice %></p>
```
Comment (blakeherrington) [inline app/views/images/show.html.erb]: There's a standard way we can implement this notice thing - let's talk about it a bit.

File: app/views/images/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+
+<p>
+  <strong>Url:</strong>
+  <span class='js-image-url'><%= @image.url %></span>
+  <br>
+  <br>
```
Comment (blakeherrington) [inline app/views/images/show.html.erb]: Instead of using two <br> tags here, we could just use a bootstrap to add some padding below the span that shows the url. You could try something like `pb-4`.

File: test/controllers/images_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    assert_raises(ActiveRecord::RecordNotFound) do
+      get image_url(-1)
+    end
+  end
+
+  def test_show__valid_record_number
+    test_url = 'http://www.maoartcorner.com/wp-content/uploads/2019/01/GlowFul.jpg'
+
+    post images_url, params: {
+      'image[url]': test_url
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Another way to structure this test is to do the below:

`image: { url: test_url }`

File: test/controllers/images_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    assert_select '.js-image-url', value: test_url
+    assert_select '.js-image', 1
+  end
+
+  def test_new__basics_intact
+    get new_image_url
+    assert_response :success
+
+    assert_select '.js-url-input', 1
+    assert_select '.js-submit-button', value: 'Create Image'
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Again, minor, but we could add an assertion here to be sure that the form is going to post to the right url - this ensures that when we click the submit button, it actually does try to hit our create action.

File: app/views/welcome/index.html.erb [hunk 15]
```diff
@@ -1 +1 @@
-<h1>Bienvenidos a localhost:3000!</h1>
+<%= link_to('Add an image +',{:action => 'new', :controller => 'images'}) %>
```
Comment (blakeherrington) [inline app/views/welcome/index.html.erb]: We could use the new hash syntax here.

`{action: new, controller: 'images' }

Or, what you more commonly see is something like this:

`link_to('Add an image', new_image_path)`

File: app/views/images/show.html.erb (hunk 11 above)
Comment (jnogales319) [reply app/views/images/show.html.erb]: Will be wiping this away as it's not part of the acceptance criteria

File: test/controllers/images_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+    assert_select '.js-image-url', value: test_url
+    assert_select '.js-image', 1
+  end
+
+  def test_new__basics_intact
+    get new_image_url
+    assert_response :success
+
+    puts body
```
Comment (blakeherrington) [inline test/controllers/images_controller_test.rb]: Oops! 😄
